Direct requestor and replier pattern samples created

gvensan opened the pull request:
Stab at creating Direct requestor an replier pattern.

gvensan committed:
Direct requestor and replier pattern samples created

gvensan committed:
Wrapping up direct messaging patterns and a helper script for simple pub/sub while testing

TamimiGitHub commented on howtos/how_to_handle_service_interruption_and_failures.py:
```diff
@@ -1,7 +1,7 @@
 
 from solace.messaging.messaging_service import MessagingService, ServiceInterruptionListener, ServiceEvent
-from solace.messaging.publisher.direct_message_publisher import PublishFailureListener
+from solace.messaging.publisher.direct_message_publisher import PublishFailureListener, FailedPublishEvent
```
We dont need to add this in the import statement since its used in the _annotation_ deceleration of the variable 
 `failed_publish_event: 'FailedPublishEvent'` That is one way to tell the compiler that `failed_publish_event` is of type `FailedPublishEvent`. You can further look into "Python annotations". The are completely optional, i.e. introduced in python 3+, however it is good practice to import them to enable code auto completion

gvensan replied:
Make sense, good to know about this auto-resolution offered by Python. For the sake of correctness and to avoid warnings - I would prefer to leave it there, but to be consistent with other references removing it.

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+## Goal: A simpe request pulblisher to publish a request with reply_to set to any topic.
```
Wrong comment for file description

gvensan replied:
Fixed

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+
+# Goal is to demonstrate receive-process-reply (processor) pattern. Processor subscribes to a request topic and upon receipt of 
+# request message carrying a string, the processor converts the payload to uppercase (for simplicity) and publishes to a 
+# designated topic.
```
Can we change the description to
> Processor subscribes to a topic and upon receipt of the message, the payload is processed (e.g. uppercased) and published to another designated topic

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+# Goal is to demonstrate receive-process-reply (processor) pattern. Processor subscribes to a request topic and upon receipt of 
+# request message carrying a string, the processor converts the payload to uppercase (for simplicity) and publishes to a 
+# designated topic.
+#   direct_requestor: 
```
`direct_processor` instead of direct_requestor

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_subscriber.py:
```diff
@@ -63,7 +63,7 @@ def on_service_interrupted(self, e: "ServiceEvent"):
 
 # Define a Topic subscriptions 
-topics = [TOPIC_PREFIX + "/dir/sub/>", TOPIC_PREFIX + "/dir/sub/v2/>", "solace/samples/>"]
+topics = [TOPIC_PREFIX + "/dir/sub/>", TOPIC_PREFIX + "/dir/sub/v2/>", "solace/samples/python/>"]
```
Can we remove the last topic subscription and just have `TOPIC_PREFIX + "/>"`

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+print(f'\nMessaging Service connected? {messaging_service.is_connected}')
+
+# Inner classes for connection error handling
+class ServiceEventHandler(ReconnectionListener, ReconnectionAttemptListener, ServiceInterruptionListener):
```
Can we move all inner classes to the top. To follow the same conventions we will have the file to start with
1. Import statements followed by
2. Environment configurations followed by
3. Constants followed by
4. Inner classes followed by
5. Main business logic followed by
6. Cleanup

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+        print("on_failed_publish")
+
+# Handle received messages
+class MessageHandlerImpl(MessageHandler):
```
Open to suggestions here: I would recommend we call this `ProcessorImpl` and assigning it as follows `direct_receiver.receive_async(ProcessorImpl())` and add descriptive comment before the start of the class. Thoughts?

TamimiGitHub replied:
I would also declare the publisher in this class as well 

```
def __init__(self, publisher):
        self.publisher = publisher
```
and then you can use it in your on_message function as follows `self.publisher.publish(destination=out_topic, message=output_msg)`

and you can pass it to the class from the main thread as follows
`direct_receiver.receive_async(MessageHandlerImpl(direct_publisher))`.

This is reflective of good practice incase the MessageHandlerImpl (or ProcessorImpl) is moved to another file and imported to the current file

gvensan replied:
Done!

gvensan replied:
I guess we are keeping all the inner classes in the same file across the samples - for consistency, let us leave it as it is. When we do the next round of house-cleaning, we will organize the inner classes in separate files and explore reusability. This is something you suggested as well :)

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+# Handle received messages
+class MessageHandlerImpl(MessageHandler):
+    def on_message(self, message: InboundMessage):
+        out_topic = Topic.of(TOPIC_PREFIX + f'/direct/processor/output')
```
Let's rename this to `publish_topic`

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+    def on_message(self, message: InboundMessage):
+        out_topic = Topic.of(TOPIC_PREFIX + f'/direct/processor/output')
+        in_topic = message.get_destination_name()
+        payload_str = message.get_payload_as_string()
```
What if the payload is not a string?

TamimiGitHub replied:
You can do the type check trick here

```
payload = message.get_payload_as_string() if message.get_payload_as_string() != None else message.get_payload_as_bytes()
        if isinstance(payload, bytearray):
            print(f"Received a message of type: {type(payload)}. Decoding to string")
            payload = payload.decode()
print(f'Received input message (body string):\n{payload}')
```

gvensan replied:
yeah. will make the change, but this could be done more elegantly either 
a) API handles necessary conversation and returns data in format as requested in the get_payload_xxx call or 
b) provide a convenience function to get the type 'xxx' and then let the user code do the necessary switch/case or if/else or even throw an error when unexpected type shows up!!

A topic for a different day... just making the change as you recommended for now.

TamimiGitHub replied:
a) API is not really meant for payload processing, i.e. what comes in goes out. type conversation is considered payload processing. And if we would to suggest that to the API implementation then I assume it would require lots of edge case management...
b) that could be, however I doubt there is a function like that in the legacy APIs. The new gen APIs were really meant to achieve feature parity with the legacy APIs. This I believe would technically be a request to the legacy API

gvensan replied:
New API can carry more useful convenience functions, which can be outside the legacy API parity aspect. API is all about making it easy/convenient for use for the developers - in this case, such a ternary (may be more if different or any type is possible) code to be repeated every time you access the payload. #justaying

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+print(f'Direct Publisher ready? {direct_publisher.is_ready()}')
+
+# Define a Topic subscriptions 
+topics = [TOPIC_PREFIX + "/direct/processor/input/>"]
```
No need for array subscription if you have only one subscription. We can just have:

```
subscribe_topic = [TopicSubscription.of(TOPIC_PREFIX + "/direct/processor/input/>")]
print(f'\nSubscribed to topic {topic.get_name()}')

# Build a Receiver with the given topics and start it
direct_receiver = messaging_service.create_direct_message_receiver_builder()\
                        .with_subscriptions(subscribe_topic)\
                        .build()
```

gvensan replied:
Done - btw, I made the subscribe topic to be specific TOPIC_PREFIX + "/direct/processor/input" (removed the wildcard, as there is no use or demonstration around that in the code)

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+# Handle received messages
+class MessageHandlerImpl(MessageHandler):
+    def on_message(self, message: InboundMessage):
+        out_topic = Topic.of(TOPIC_PREFIX + f'/direct/processor/output')
```
can we change var name to `publish_topic`

gvensan replied:
Done - also changed the in_topic -> subscribe_topic

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+direct_receiver.start()
+print(f'Direct Subscriber is running? {direct_receiver.is_running()}')
+
+# Prepare outbound message payload and body
```
This block is not needed, `msg_builder` is not used

TamimiGitHub replied:
I realized you are using it in the message handler. I suggest moving there

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+
+        output_msg = msg_builder \
+                        .with_application_message_id(message.get_application_message_id())\
+                        .build(f'{processed_payload_str}')
```
I would recommend also adding properties to the published message 
`.with_property("payload", "processed") \`

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+                .with_property("application", "samples") \
+                .with_property("language", "Python") \
+
+count = 1
```
count not used

gvensan replied:
Removed

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+# request message carrying a string, the processor converts the payload to uppercase (for simplicity) and publishes to a 
+# designated topic.
+#   direct_requestor: 
+#       subscribes to request topic 'solace/samples/python/direct/processor/input/{count}
```
`subscribes to request topic 'solace/samples/python/direct/processor/input/>`

gvensan replied:
Changed the requested topic without any wildcards as there is no function demonstrated on the topic variable.

subscribes to request topic  'solace/samples/python/direct/processor/input'

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+        in_topic = message.get_destination_name()
+        payload_str = message.get_payload_as_string()
+        print(f'Received input message (body string):\n{payload_str}')
+        print(f'Message dump: {message.solace_message.get_message_dump()} \n')
```
to dump the message you can just 
`print(f'Message dump: {message} \n')`

gvensan replied:
Removing the dump print feels redundant.

TamimiGitHub commented:
Received the following error when running 
```
2022-02-22 18:19:21,227 [WARNING] solace.messaging.publisher: [_solace_utilities.py:227]  Invalid datatype. Expected type: [<class 'str'>], but actual [<class 'NoneType'>]
2022-02-22 18:19:21,227 [WARNING] solace.messaging.receiver: [_direct_message_receiver.py:104]  [_DirectMessageReceiverThread Id: 0x105d89340] Failed to dispatch the callback Invalid datatype. Expected type: [<class 'str'>], but actual [<class 'NoneType'>]
```

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+from solace.messaging.resources.topic_subscription import TopicSubscription
+from solace.messaging.receiver.message_receiver import MessageHandler, InboundMessage
+
+
```
Extra new line

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+        print(f'Published output message (body):\n{processed_payload_str}')
+
+        output_msg = msg_builder \
+                        .with_application_message_id(message.get_application_message_id())\
```
As per the [docs](https://docs.solace.com/API-Developer-Online-Ref-Documentation/python/source/rst/solace.messaging.html?#solace.messaging.message.Message.get_application_message_id), `get_application_message_id()` 
>Returns Sender application identifier if set by message publisher, or None/empty if not set.

If the sender does not set the application message ID you will get the following warning when attempting to build a message builder 
>[WARNING] solace.messaging.publisher: [_solace_utilities.py:227]  Invalid datatype. Expected type: [<class 'str'>], but actual [<class 'NoneType'>]

gvensan replied:
Checked for get_application_message_id() is not NONE, then set it on the published message.

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+        print(f'>>>>>>>>>>>>>>>>>>>>>>>>>>>>')
+        print(f'Published output message on {out_topic}')
+        print(f'Published output message (body):\n{processed_payload_str}')
+        print(f'----------------------------')
```
Here is one way to implement the message handler

```
# Handle received messages
class MessageHandlerImpl(MessageHandler):
    def __init__(self, publisher):
        self.publisher = publisher

    def on_message(self, message: InboundMessage):
        payload = message.get_payload_as_string() if message.get_payload_as_string() != None else message.get_payload_as_bytes()
        if isinstance(payload, bytearray):
            print(f"Received a message of type: {type(payload)}. Decoding to string")
            payload = payload.decode()
        
        publish_topic = Topic.of(TOPIC_PREFIX + f'/direct/processor/output')
        print(f'Received input message:{payload}')

        processed_payload_str = payload.upper()
        print(f"Processed payload: {processed_payload_str}")
        print(f"Republishing processed message on: {publish_topic.get_name()}")

        # Build the outbound message
        outbound_msg = messaging_service.message_builder() \
                .with_property("application", "processor") \
                .with_property("language", "Python") \
                .build(processed_payload_str)\
        
        # Publish the processed message on the publish_topic
        self.publisher.publish(destination=publish_topic, message=outbound_msg)
```
what do you think?

gvensan replied:
I don't understand what is done in this line:

>         payload = message.get_payload_as_string() if message.get_payload_as_string() != None else message.get_payload_as_bytes()

Please review the changes I made based on the previous comments - it is as good as what you have shown here. If anything missing, we can revise it for sure.

TamimiGitHub replied:
This is pretty much how a ternary operator is implemented in python. It's common practice to adopt a ternary operator if there is a one liner if else body. This lines basically reads:
> Set the payload variable to the string representation of the message payload if its not None. otherwise just set it to the byte representation

This is how a ternary operator is set in [Java](https://www.geeksforgeeks.org/java-ternary-operator-with-examples) and [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)

TamimiGitHub replied:
I propose to keep it as is, and I'll make another PR to change the implementation of all the on_message methods for the message handlers in all the patterns in one PR

gvensan replied:
Got it... it is a weird ternary operator tho :)

TamimiGitHub reviewed (changes requested):
I reviewed the direct processor, great first stab at it! Just a couple of comments and rearranging based on conventions. Will review the remaining samples tomorrow 👍

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+## Goal: A simpe request pulblisher to publish a request with reply_to set to any topic.
```
Typo: simple, publisher

TamimiGitHub replied:
I would also mention that this is a blocking requester

gvensan replied:
Removed that block of comments - just sticking with convention fo 1-liner at the top of the file like other samples.

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+import os
+import platform
+import time
+import calendar;
```
extra semicolon at the end

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+print('\nSend a KeyboardInterrupt to stop publishing')
+try: 
+    print(f'============================')
+    topic = Topic.of(TOPIC_PREFIX + '/direct/hello/'  + f'{unique_name}')
```
`topic = Topic.of(TOPIC_PREFIX + '/direct/request/'  + f'{unique_name}')`

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+messaging_service.add_service_interruption_listener(service_handler)
+
+# Create a direct message requestor and register the error handler
+direct_requestor_blocking = messaging_service.request_reply() \
```
Can we check with the API team if there is a way to explicitly set the reply_to topic? Instead of having the API dynamically setting a #P2P topic for the reply to. In the legacy API, there is a way to set the reply to header prop of the message prior to sending it

setReplyTo(Destination destination)
--


[setReplyTo](https://docs.solace.com/API-Developer-Online-Ref-Documentation/java/com/solacesystems/jcsmp/XMLMessage.html#setReplyTo-com.solacesystems.jcsmp.Destination-)([Destination](https://docs.solace.com/API-Developer-Online-Ref-Documentation/java/com/solacesystems/jcsmp/Destination.html) destination)

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+
+        response = direct_requestor_blocking.publish_await_response(request_message=outbound_msg, \
+                                                                    request_destination=topic, \
+                                                                    reply_timeout=10000)
```
let's change the timeout to `3000` and follow the same convention in https://github.com/SolaceSamples/solace-samples-java-jcsmp/blob/master/src/main/java/com/solace/samples/jcsmp/patterns/DirectRequestorBlocking.java#L45

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+print(f'\nDirect Requestor ready? {direct_requestor_blocking.is_ready()}')
+
+# Prepare outbound message payload and body
+message_body = "Hi, My name is " + f'{name}'
```
can we change the message body to be "This is request message from {name}"

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+    print('\nTerminating Requestor')
+    direct_requestor_blocking.terminate()
+    print('\nDisconnecting Messaging Service')
+    messaging_service.disconnect()
```
minor: newline

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,142 @@
+from solace.messaging.publisher.direct_message_publisher import PublishFailureListener, FailedPublishEvent
+from solace.messaging.resources.topic_subscription import TopicSubscription
+from solace.messaging.receiver.message_receiver import MessageHandler, InboundMessage
+
```
I would recommend importing the RequestReplyMessagePublisher
`from solace.messaging.publisher.request_reply_message_publisher import RequestReplyMessagePublisher`

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+            ServiceInterruptionListener, RetryStrategy, ServiceEvent
+from solace.messaging.resources.topic import Topic
+from solace.messaging.errors.pubsubplus_client_error import PubSubPlusClientError
+
```
I would recommend importing the RequestReplyMessagePublisher
`from solace.messaging.publisher.request_reply_message_publisher import RequestReplyMessagePublisher`

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor_blocking.py:
```diff
@@ -0,0 +1,121 @@
+messaging_service.add_service_interruption_listener(service_handler)
+
+# Create a direct message requestor and register the error handler
+direct_requestor_blocking = messaging_service.request_reply() \
```
I would recommend specifying the variable type with an annotation
```
direct_requestor_blocking: RequestReplyMessagePublisher = messaging_service.request_reply() \
                                    .create_request_reply_message_publisher_builder() \
                                    .build()
```
That way you can access the `RequestReplyMessagePublisher` class methods easier from your IDE

gvensan replied:
Done

TamimiGitHub commented:
Out of curiosity, what is the [helper_pupsub.py](https://github.com/SolaceSamples/solace-samples-python/pull/21/files#diff-b1ad74505e06eb76af045731f32a43016e9e3ba54ac9555597c816abcdd4267b)? Looks interesting but could be in a separate branch/PR

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+import os
+import platform
+import time
+import calendar;
```
Similar comment in the blocking requestor: semicolon

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+            ServiceInterruptionListener, RetryStrategy, ServiceEvent
+from solace.messaging.resources.topic import Topic
+from solace.messaging.errors.pubsubplus_client_error import PubSubPlusClientError
+
```
Similar comment in the blocking requestor: import the RequestReplyMessagePublisher

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+messaging_service.add_service_interruption_listener(service_handler)
+
+# Create a direct message requestor and register the error handler
+direct_requestor = messaging_service.request_reply() \
```
Similar comment in the blocking requestor: add annotation to specify `: RequestReplyMessagePublisher`

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+print(f'\nDirect Requestor ready? {direct_requestor.is_ready()}')
+
+# Prepare outbound message payload and body
+message_body = "Hi, My name is " + f'{name}'
```
Similar comment in the blocking requestor: change message body

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+print('\nSend a KeyboardInterrupt to stop publishing')
+try:
+    print(f'============================')
+    topic = Topic.of(TOPIC_PREFIX + '/direct/hello/'  + f'{unique_name}')
```
Similar comment in the blocking requestor: change topic to match samples conventions

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,121 @@
+        print(f'Publishing message:\n{outbound_msg}')
+        publish_async = direct_requestor.publish(request_message=outbound_msg, \
+                                                request_destination=topic,
+                                                reply_timeout=10000)
```
Similar comment in the blocking requestor: timeout

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+    "solace.messaging.authentication.scheme.basic.password": os.environ.get('SOLACE_PASSWORD') or ""
+    }
+
+request_topic = TOPIC_PREFIX + '/direct/hello/>'
```
`'/direct/requestor/>'`

gvensan replied:
Done - used '/direct/request/>'

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+import os
+import platform
+import time
+import calendar;
```
semicolon

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+            ServiceInterruptionListener, RetryStrategy, ServiceEvent
+from solace.messaging.resources.topic import Topic
+from solace.messaging.resources.topic_subscription import TopicSubscription
+from solace.messaging.receiver.request_reply_message_receiver import RequestMessageHandler, InboundMessage, Replier
```
Import `RequestReplyMessageReceiver` as well from the request_reply_message_receiver

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+messaging_service.add_reconnection_attempt_listener(service_handler)
+messaging_service.add_service_interruption_listener(service_handler)
+
+# Create a direct message replier and register the error handler
```
let's remove `and register the error handler` from the comment since we are not registering any errors to the replier

gvensan replied:
Done

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+        print(f'----------------------------')
+        print(f'Received request:\n{request}')
+
+        message_id = request.get_application_message_id();
```
assuming the message_id is set by the requestor

gvensan replied:
Yes - this sample sets the same message-id on the response as set on the request!

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+# Handle received messages
+class RequestMessageHandlerImpl(RequestMessageHandler):
+    def on_message(self, request: InboundMessage, replier: Replier):
+        name = request.get_payload_as_string().split("My name is ")[1]
```
This will return the following error if "my name is" is not found in the payload 
```
list index out of range
```
to avoid this I would recommend checking if 'my name is "  is a substring of the request payload using the python `if substring in string` notation
```
response_body = f'Greetings {request_payload_str.split("This is request message from ")[1]}!' if "This is request message from " in request_payload_str else 'Response body'

outbound_msg = self.message_builder \
                            .with_application_message_id(f'{message_id}')\
                            .build(f'{response_body}')

```

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+# Handle received messages
+class RequestMessageHandlerImpl(RequestMessageHandler):
+    def on_message(self, request: InboundMessage, replier: Replier):
+        name = request.get_payload_as_string().split("My name is ")[1]
```
This also assumes that the request payload is string

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+        else:
+            print(f'Invalid request, reply_to not set')    
+        print(f'============================')
+
```
Made some modifications to the RequestMessageHandlerImpl, what are your thoughts?

```
class RequestMessageHandlerImpl(RequestMessageHandler):
    def __init__(self, message_builder):
        self.message_builder = message_builder

    def on_message(self, request: InboundMessage, replier: Replier):
        request_payload_str = request.get_payload_as_string() if request.get_payload_as_string() != None else request.get_payload_as_bytes()
        if isinstance(request_payload_str, bytearray):
            print(f"Received a message of type: {type(request_payload_str)}. Decoding to string")
            request_payload_str = request_payload_str.decode()

        response_body = f'Greetings {request_payload_str.split("My name is ")[1]}!' if "My name is " in request_payload_str else "Response body"

        print(f'Received request: {request_payload_str}')
        # print(f'Received request message dump: {request}')

        message_id = request.get_application_message_id()
        if replier is not None:
            outbound_msg = self.message_builder \
                            .with_application_message_id(f'{message_id}')\
                            .build(f'{response_body}')
            replier.reply(outbound_msg)
            print(f'>>>>>>>>>>>>>>>>>>>>>>>>>>>>')
            print(f'Replied with response (body): {outbound_msg.get_payload_as_string()}')
            print(f'----------------------------')
            print(f'Response Message dump: {outbound_msg}')
        else:
            print(f'Invalid request, reply_to not set')    
        print(f'============================')
```

Note that this means that the RequestMessageHandlerImpl is called as follows
```
direct_replier.receive_async(RequestMessageHandlerImpl(outbound_msg_builder))
```

gvensan replied:
Incorporated.

TamimiGitHub commented on patterns/direct_replier.py:
```diff
@@ -0,0 +1,122 @@
+messaging_service.add_service_interruption_listener(service_handler)
+
+# Create a direct message replier and register the error handler
+direct_replier = messaging_service.request_reply() \
```
You can use the annotation in python as follows 
`direct_replier: RequestReplyMessageReceiver = ....`

gvensan replied:
Done!

gvensan committed:
a) Incorporated PR comments/feedback from Tamimi
b) A working version of Guaranteed Processor is checked in for review
c) Bit more cleanup to make the code look consistent with other samples

gvensan commented:
> Out of curiosity, what is the [helper_pupsub.py](https://github.com/SolaceSamples/solace-samples-python/pull/21/files#diff-b1ad74505e06eb76af045731f32a43016e9e3ba54ac9555597c816abcdd4267b)? Looks interesting but could be in a separate branch/PR

The 'helper_pubsub.py' is a general-purpose quick publish/subscribe utility that you can run from the command line. The Try-me of manager has issues like not being able to set the content type and act on multiple events. It was meant for my testing help - not to be checked in. I will delete it...

Will add it to a new folder 'utilities' as we discussed.

gvensan committed:
Removing the helper_pubsub.py, meant to be for local testing.

TamimiGitHub commented on patterns/guaranteed_processor.py:
```diff
@@ -0,0 +1,163 @@
+import time
+import threading
+
+# Import Solace Python  API modules from the solace package
```
Extra space in `Solace Python  API `

gvensan replied:
:)

TamimiGitHub commented on patterns/guaranteed_processor.py:
```diff
@@ -0,0 +1,163 @@
+
+if platform.uname().system == 'Windows': os.environ["PYTHONUNBUFFERED"] = "1" # Disable stdout buffer 
+
+lock = threading.Lock() # lock object that is not owned by any thread. Used for synchronization and counting the 
```
hmm why are we using locks here?

TamimiGitHub commented on patterns/guaranteed_processor.py:
```diff
@@ -0,0 +1,163 @@
+    def receipt_count(self):
+        return self._receipt_count
+
+    def on_publish_receipt(self, publish_receipt: 'PublishReceipt'):
```
We should do something with the publish_receipt. So check if the publish_receipt.exception is != None then print it out.
https://docs.solace.com/API-Developer-Online-Ref-Documentation/python/source/rst/solace.messaging.publisher.html?#solace.messaging.publisher.persistent_message_publisher.PublishReceipt

TamimiGitHub replied:
I realized we're not handling the publihs receipt in the guaranteed publisher sample as well 
https://github.com/SolaceSamples/solace-samples-python/issues/22
So we can do it after this PR is merged

gvensan replied:
Cool.

TamimiGitHub commented on patterns/guaranteed_processor.py:
```diff
@@ -0,0 +1,163 @@
+        self.publisher = publisher
+
+        # Prepare outbound message payload and body
+        self.msg_builder = messaging_service.message_builder() \
```
We should avoid using global variables in constructors of classes and pass the variables as parameters. The same way you passed the publisher variable, you can also pass the messaging_service as a parameter and then create the msg_builder in the constructor of the class as follows

```
def __init__(self, publisher, messaging_service):
        self.publisher = publisher
        self.msg_builder = messaging_service.message_builder() \
                                .with_application_message_id("sample_id") \
                                .with_property("application", "samples") \
                                .with_property("payload", "processed") \
                                .with_property("language", "Python") \

```
and before you call the `ProcessorImpl` class, you can pass in the messaging_service to it as follows

```
persistent_receiver.receive_async(ProcessorImpl(publisher, messaging_service))
```

gvensan replied:
Is this a python thing, I am learning :) Will make changes...

TamimiGitHub commented on patterns/direct_requestor.py:
```diff
@@ -0,0 +1,118 @@
+print(f'\nDirect Requestor ready? {direct_requestor.is_ready()}')
+
+# Prepare outbound message payload and body
+message_body = "This is request message from " + f'{name}'
```
`message_body = f'This is request message from {name}' `

gvensan replied:
Done!

TamimiGitHub commented on patterns/direct_processor.py:
```diff
@@ -0,0 +1,147 @@
+        print("on_failed_publish")
+
+# Inner class to handle received messages
+class ProcessorImpl(MessageHandler):
```
Similar comment to the `ProcessorImpl` class in the guaranteed processor: pass the messaging service or the outbound message to the class as a parameter

gvensan replied:
Done!

TamimiGitHub commented on patterns/guaranteed_processor.py:
```diff
@@ -0,0 +1,163 @@
+    try: 
+        while True:
+            time.sleep(1)
+    except KeyboardInterrupt:
```
We're not handling the exception if the queue does not exist 
```
from solace.messaging.errors.pubsubplus_client_error import PubSubPlusClientError


# Handle API exception 
except PubSubPlusClientError as exception:
  print(f'\nMake sure queue {queue_name} exists on broker!')
```
See guaranteed subscriber for more details

gvensan replied:
Done!

TamimiGitHub commented:
Looking great after the changes @gvensan ! I have resolved the [conversations](https://github.com/SolaceSamples/solace-samples-python/pull/21/files) that were incorporated and left the couple that are yet to be resolved after you latest merge 👍

gvensan committed:
- Incorporated pending (and new) feedback on the PR
- Moved the helper_pubsub.py to utilities directory within patterns folder

TamimiGitHub commented:
LGTM!! We'll go ahead and merge :D
